### ISSUE TITLE
Adding a `resources:ram:1234` tag to a test doesn't work and causes a crash

### ISSUE THREAD

**Issue by jmmv** [bazelbuild/bazel#19572]
### Description of the bug:

I'm in a situation where I need to tag some tests as requiring lots of memory so that Bazel doesn't schedule more than one at once. Using `size=enormous` isn't sufficient because the built-in 800MB setting is too low and cannot be changed. Using `exclusive` is too heavy handed because it prevents all parallelism.

Skimming through the code in `TestTargetProperties.java`, I found support for a promising `resources:<resourcename>:<amount>` tag so I tried to tag the tests with `resources:ram:1234`. Unfortunately, doing so made Bazel crash immediately with:

```
FATAL: bazel crashed due to an internal error. Printing stack trace:
java.lang.RuntimeException: Unrecoverable error while evaluating node 'UnshareableActionLookupData{actionLookupKey=ConfiguredTargetKey{label=//foo:bar, config=BuildConfigurationKey[67e98e6b54e7bc1ffbd4d497c3b49020e0550b284d05c428b57c7ea699b61bca]}, actionIndex=8}' (requested by nodes 'TestCompletionKey{configuredTargetKey=ConfiguredTargetKey{label=//foo:bar, config=BuildConfigurationKey[67e98e6b54e7bc1ffbd4d497c3b49020e0550b284d05c428b57c7ea699b61bca]}, topLevelArtifactContext=com.google.devtools.build.lib.analysis.TopLevelArtifactContext@90904c3b, exclusiveTesting=true}')
        at com.google.devtools.build.skyframe.AbstractParallelEvaluator$Evaluate.run(AbstractParallelEvaluator.java:633)
        at com.google.devtools.build.lib.concurrent.AbstractQueueVisitor$WrappedRunnable.run(AbstractQueueVisitor.java:365)
        at java.base/java.util.concurrent.ForkJoinTask$AdaptedRunnableAction.exec(Unknown Source)
        at java.base/java.util.concurrent.ForkJoinTask.doExec(Unknown Source)
        at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(Unknown Source)
        at java.base/java.util.concurrent.ForkJoinPool.scan(Unknown Source)
        at java.base/java.util.concurrent.ForkJoinPool.runWorker(Unknown Source)
        at java.base/java.util.concurrent.ForkJoinWorkerThread.run(Unknown Source)
Caused by: java.util.NoSuchElementException: Resource ram is not tracked in this resource set.
        at com.google.devtools.build.lib.actions.ResourceManager.assertExtraResourcesTracked(ResourceManager.java:499)
        at com.google.devtools.build.lib.actions.ResourceManager.areResourcesAvailable(ResourceManager.java:540)
        at com.google.devtools.build.lib.actions.ResourceManager.acquire(ResourceManager.java:384)
        at com.google.devtools.build.lib.actions.ResourceManager.acquireResources(ResourceManager.java:256)
        at com.google.devtools.build.lib.exec.local.LocalSpawnRunner.exec(LocalSpawnRunner.java:161)
        at com.google.devtools.build.lib.sandbox.SandboxModule$SandboxFallbackSpawnRunner.exec(SandboxModule.java:499)
        at com.google.devtools.build.lib.exec.SpawnRunner.execAsync(SpawnRunner.java:301)
        at com.google.devtools.build.lib.exec.AbstractSpawnStrategy.exec(AbstractSpawnStrategy.java:152)
        at com.google.devtools.build.lib.dynamic.LocalBranch.runSpawnLocally(LocalBranch.java:128)
        at com.google.devtools.build.lib.dynamic.LocalBranch.runLocally(LocalBranch.java:94)
        at com.google.devtools.build.lib.dynamic.DynamicSpawnStrategy.maybeExecuteNonDynamically(DynamicSpawnStrategy.java:369)
        at com.google.devtools.build.lib.dynamic.DynamicSpawnStrategy.exec(DynamicSpawnStrategy.java:193)
        at com.google.devtools.build.lib.actions.SpawnStrategy.beginExecution(SpawnStrategy.java:47)
        at com.google.devtools.build.lib.exec.SpawnStrategyResolver.beginExecution(SpawnStrategyResolver.java:64)
        at com.google.devtools.build.lib.exec.StandaloneTestStrategy.beginTestAttempt(StandaloneTestStrategy.java:323)
        at com.google.devtools.build.lib.exec.StandaloneTestStrategy$StandaloneTestRunnerSpawn.beginExecution(StandaloneTestStrategy.java:599)
        at com.google.devtools.build.lib.analysis.test.TestRunnerAction.beginIfNotCancelled(TestRunnerAction.java:995)
        at com.google.devtools.build.lib.analysis.test.TestRunnerAction.beginExecution(TestRunnerAction.java:962)
        at com.google.devtools.build.lib.analysis.test.TestRunnerAction.execute(TestRunnerAction.java:1019)
        at com.google.devtools.build.lib.analysis.test.TestRunnerAction.execute(TestRunnerAction.java:1011)
        at com.google.devtools.build.lib.skyframe.SkyframeActionExecutor$5.execute(SkyframeActionExecutor.java:961)
        at com.google.devtools.build.lib.skyframe.SkyframeActionExecutor$ActionRunner.continueAction(SkyframeActionExecutor.java:1128)
        at com.google.devtools.build.lib.skyframe.SkyframeActionExecutor$ActionRunner.run(SkyframeActionExecutor.java:1086)
        at com.google.devtools.build.lib.skyframe.ActionExecutionState.runStateMachine(ActionExecutionState.java:160)
        at com.google.devtools.build.lib.skyframe.ActionExecutionState.getResultOrDependOnFuture(ActionExecutionState.java:93)
        at com.google.devtools.build.lib.skyframe.SkyframeActionExecutor.executeAction(SkyframeActionExecutor.java:519)
        at com.google.devtools.build.lib.skyframe.ActionExecutionFunction.checkCacheAndExecuteIfNeeded(ActionExecutionFunction.java:827)
        at com.google.devtools.build.lib.skyframe.ActionExecutionFunction.computeInternal(ActionExecutionFunction.java:323)
        at com.google.devtools.build.lib.skyframe.ActionExecutionFunction.compute(ActionExecutionFunction.java:161)
        at com.google.devtools.build.skyframe.AbstractParallelEvaluator$Evaluate.run(AbstractParallelEvaluator.java:562)
        ... 7 more
```

Looking further at the code, it seems that this tag only works to specify the CPU needs of a test (and `resources:cpu:X` is equivalent to `cpu:X`). I also found the `--local_extra_resources` flag, but I have no idea how this plays into the game as it doesn't seem to impact the default resource sets and the flag is not well-documented.

I would like `resources:ram:X` to override the RAM requirements of a test, ignoring whatever the `size` property decides. I do not want to create a new dimension of resources, as `--local_extra_resources` seems to do per https://github.com/bazelbuild/bazel/issues/18153. And Bazel should not crash.

### Which category does this issue belong to?

Local Execution

### What's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

Add a `resources:ram:1` tag to a test, try to run the test, and see Bazel crash.

### Which operating system are you running Bazel on?

Linux

### What is the output of `bazel info release`?

release 6.3.2

**Comment by DavidANeil** [comment]
I believe the tag is `resources:memory:1234`, which seems to be working correctly for me.

I wonder if it is worth having a special mapping of the resource name `ram` and `memory`, since the old flag was `--local_ram_resources=1234`, but the new way will be `--local_resources=memory=1234`.  Seems like a lot of people are going to get bitten by this.

Perhaps a better solution:
Using a `resources:` tag for an unknown resource should display a calm error message and exit, instead of crashing the server process.  The error message should include the available resources, so it would be quite obvious that `memory` is one of the options.

**Comment by tjgq** [comment]
We would welcome an external contribution here.

**Comment by AlessandroPatti** [comment]
@tjgq submitted a patch [here](https://github.com/bazelbuild/bazel/pull/23541)

**Comment by iancha1992** [comment]
A fix for this issue has been included in [Bazel 7.4.0 RC1](https://github.com/bazelbuild/bazel/releases/tag/7.4.0rc1). Please test out the release candidate and report any issues as soon as possible.
If you're using Bazelisk, you can point to the latest RC by setting `USE_BAZEL_VERSION=7.4.0rc1`. Thanks!
